### ISSUE TITLE
Store class definitions in database

### ISSUE THREAD

**Issue by BlackbitDevs** [pimcore/pimcore#14046]
Nowadays it is possible to store almost all settings, which can be changed in the Pimcore backend, in the database. This allows client users to change a lot of things without having to know how to get the changes to Git - despite the fact that in SaaS solutions it is easier to have all settings in the database. But there is one exception: class definitions.

In https://github.com/pimcore/pimcore/issues/8156#issuecomment-804842638 @brusch said
> I honestly don't want to see the class definitions stored in DB

What is the problem about storing class definitions in the database? 
I mean, Pimcore does the same with classification store definitions.

Of course it may not be optimal to store the definitions as serialized PHP but of course this could also be refactored.

**Comment by dpfaffenbauer** [comment]
Storing ClassDefinitions in the DB also means re-creating the PHP Classes, which in Terms of Containerized workflows might not work to readonly filesystems. I am also against doing that.

**Comment by BlackbitDevs** [comment]
I have some problems to understand you, @dpfaffenbauer  ;)

> Storing ClassDefinitions in the DB also means re-creating the PHP Classes

Which PHP classes do you mean? The data object classes? Don't they have to be generated anyway if the definitions change?

> which in Terms of Containerized workflows might not work to readonly filesystems

Can you explain this setup? Does it mean that the generated data object classes get generated somewhere else than in the system which uses them?

**Comment by dpfaffenbauer** [comment]
> Which PHP classes do you mean? The data object classes? Don't they have to be generated anyway if the definitions change?

Yes, but currently the most used workflow is: Develop and define the class-definitions locally, this creates the php classes locally too. On deployment (or docker build), use `pimcore:build:classes` or `pimcore:deployment:classes-rebuild` to create them.

> Can you explain this setup? Does it mean that the generated data object classes get generated somewhere else than in the system which uses them?

The PHP Classes get generated on docker build with `pimcore:build:classes`. For security reasons, the whole filesystem is readonly.

**Comment by BlackbitDevs** [comment]
> Yes, but currently the most used workflow is: Develop and define the class-definitions locally, this creates the php classes locally too. On deployment (or docker build), use pimcore:build:classes or pimcore:deployment:classes-rebuild to create them.

You are right but how is this different from thumbnail definitions for example? Or classification store fields? If you want to change those things locally and then deploy them, you are never able to use database storage for the definitions (except when you deploy the changes with database migrations). Nevertheless those things can be stored in the database.

> The PHP Classes get generated on docker build with pimcore:build:classes. For security reasons, the whole filesystem is readonly.

Ok, but in this case you will not be able to change the classes in the Pimcore backend anyway, right?


I know that there different deployment strategies and I do not want to remove the possibility to store the class definitions in files. But imagine SaaS setups: Multiple projects share the complete same Pimcore files - the only difference is that they use different databases (which contains different thumbnail definitions, reports - and in optimal case also class definitions).

**Comment by dpfaffenbauer** [comment]
> You are right but how is this different from thumbnail definitions for example? Or classification store fields? If you want to change those things locally and then deploy them, you are never able to use database storage for the definitions (except when you deploy the changes with database migrations). Nevertheless those things can be stored in the database.

True, in my setups that isn't possible.

> I know that there different deployment strategies and I do not want to remove the possibility to store the class definitions in files. But imagine SaaS setups: Multiple projects share the complete same Pimcore files - the only difference is that they use different databases (which contains different thumbnail definitions, reports - and in optimal case also class definitions).

Pimcore is not build for SaaS.

**Comment by BlackbitDevs** [comment]
> Pimcore is not build for SaaS.

What is https://pimcore.com/de/cloud then? This is exactly the fixed setup which I am speaking about. You cannot add any frontend application or plugins in this setup - because all clients get the same Pimcore setup - which they can adjust of course to their needs. But not on filesystem level but only on database. I think for this Pimcore implemented some things that it still works with the file-based class definitions but imagine how much easier it could be if ALL client-specific settings in this case would be stored in the database.

**Comment by dkarlovi** [comment]
I think this is an interesting topic, by default I agree with @dpfaffenbauer: as it stands, Pimcore isn't built for this and the safe way forward is treating classes as prebuilt, read-only things. You don't touch class definitions in prod.

But, having built a similar (non-Pimcore) system from scratch: Pimcore **could** be built for this, you could have a multi-tenant Pimcore install, but this is something Pimcore would need to fully embrace, it's not just about putting class definitions in a database, there's a lot of questions which would need to be answered here, for example:
1. where do we put the actual classes?
2. when do we build them?
3. what happens in a HA environment where you have multiple containers / servers serving the same Pimcore instance, which one builds it? If all, how do they coordinate it?
4. do we still build them as is or are they in some way different (say, ad-hoc anonymous in memory classes?)

The classificaton store is a good measure how complex this starts to get and the implementation there gives a good estimate what issues this approach would have. Again, I think this would be doable, but it would require a lot of non-trivial work.
